smelter-core/src/pipeline/rtp/rtp_input/jitter_buffer.rs

```diff
@@ -114,42 +114,51 @@ impl RtpJitterBuffer {
-                match self.mode {
-                    RtpJitterBufferMode::Fixed(duration) => {
-                        // if input is required or offset is set, we can assume that we can wait
-                        // a while, but it should not depend on queue clock
-                        if first_packet.received_at.elapsed() < duration {
-                            return None;
-                        }
-                    }
-                    RtpJitterBufferMode::QueueBased => {
-                        let lowest_pts = self.packets.values().map(|packet| packet.pts).min()?;
-
-                        // TODO: if lowest pts is not first it means that we have B-frames
-                        //
-                        // It would be safer to use value based on index than constant, in the worst
-                        // case scenario this could be 16 frames that needs to decoded in that time
-                        let should_pop = lowest_pts + self.input_buffer.size()
-                            < self.queue_sync_point.elapsed() + MIN_DECODE_TIME;
-                        if !should_pop {
-                            return None;
-                        }
-                    }
-                }
-                (self.on_stats_event)(RtpJitterBufferStatsEvent::RtpPacketLost(
-                    first_seq_num.saturating_sub(previous_seq_num),
-                ));
+        if self.next_seq_num == Some(*first_seq_num) {
+            return self.pop();
+        }
+
+        let wait_for_next_packet = match self.mode {
+            RtpJitterBufferMode::Fixed(duration) => {
+                // if input is required or offset is set, we can assume that we can wait
+                // a while, but it should not depend on queue clock
+                first_packet.received_at.elapsed() < duration
+            }
+            RtpJitterBufferMode::QueueBased => {
+                let lowest_pts = self.packets.values().map(|packet| packet.pts).min()?;
+
+                // TODO: if lowest pts is not first it means that we have B-frames
+                //
+                // It would be safer to use value based on index than constant, in the worst
+                // case scenario this could be 16 frames that needs to decoded in that time
+                let next_pts = lowest_pts + self.input_buffer.size();
+                next_pts > self.queue_sync_point.elapsed() + MIN_DECODE_TIME
+            }
+            RtpJitterBufferMode::Disabled => {
+                return self.pop();
             }
         };
+        if wait_for_next_packet {
+            return None;
+        }
+
+        match self.next_seq_num {
+            Some(next) => {
+                (self.on_stats_event)(RtpJitterBufferStatsEvent::RtpPacketLost);
+                self.next_seq_num = Some(next + 1);
+                Some(RtpInputEvent::LostPacket)
+            }
+            None => {
+                // first packet
+                self.pop()
```

Comment:
The first packet will only be popped now when we are running out of time, previouslly first packet was popped immediately after receiving it. However, that risked that if the initial packet was reorded and was the keyframe we would never request new one